util: Use consteval checked format string in FatalErrorf, LogConnectFailure

maflcko opened the pull request:
The `test/lint/lint-format-strings.py` was designed to count the number of format specifiers and assert that they are equal to the number of parameters passed to the format function. The goal seems reasonable, but the implementation has many problems:

* It is written in Python, meaning that C++ code can not be parsed correctly. Currently it relies on brittle regex and string parsing.
* Apart from the parsing errors, there are also many logic errors. For example, `count_format_specifiers` allows a mix of positional specifiers and non-positional specifiers, which can lead to runtime format bugs. Also, `count_format_specifiers` silently skipped over "special" format specifiers, which are valid in tinyformat, which again can lead to runtime format bugs being undetected.
* The brittle logic has a history of breaking in pull requests that are otherwise fine. This causes the CI to fail and the pull request being blocked from progress until the bug in the linter is fixed, or the code is rewritten to work around the bug.
* It is only run in the CI, or when the developer invokes the script. It would be better if the developer got the error message at compile-time, directly when writing the code.

Fix all issues by using a `consteval` checked format string in `FatalErrorf` and `LogConnectFailure`.

This is the first step toward https://github.com/bitcoin/bitcoin/issues/30530 and a follow-up will apply the approach to the other places.

Commit:
util: Add ConstevalFormatString

The type is used to wrap a format string once it has been compile-time
checked to contain the right number of format specifiers.

Commit:
util: Use compile-time check for FatalErrorf

TheCharlatan commented on src/util/string.h:
```diff
@@ -17,6 +17,28 @@
+            }
+            if (perc) begin_f = true;
+        }
+        assert(num_params == count);
```
I think it would be good practice to add a comment directly behind these asserts in consteval functions explaining what the error is, so we get better error messages e.g.
```
assert(num_params == count); // Format arguments have to match the number of format specifiers in the string.
```

TheCharlatan commented on src/index/base.cpp:
```diff
@@ -27,9 +27,9 @@ constexpr auto SYNC_LOG_INTERVAL{30s};
 
 template <typename... Args>
-void BaseIndex::FatalErrorf(const char* fmt, const Args&... args)
+void BaseIndex::FatalErrorf(util::ConstevalFormatString<sizeof...(Args)> fmt, const Args&... args)
```
Nit: Add include for `util/string.h`

sedited reviewed (approved):
ACK fa1f64d15e75137a6d4b469e7de6e1be0fda762a

l0rinc commented on src/util/string.cpp:
```diff
@@ -8,6 +8,25 @@
 namespace util {
+
+// Compile-time sanity checks
+static_assert([] {
```
This would pass if we forgot to add an exception, we need to test the negatives as well. This is the reason for my NACK

l0rinc commented on src/util/string.h:
```diff
@@ -17,6 +17,28 @@
+    consteval ConstevalFormatString(const char* str) : fmt{str} { CheckNumFormatSpecifiers(fmt); }
+    consteval static void CheckNumFormatSpecifiers(std::string_view str)
+    {
+        int count{0};
+        bool begin_f{false};
+        for (char c : str) {
+            bool perc{c == '%'};
+            if (begin_f) {
+                count += !perc;
+                begin_f = false;
+                continue;
+            }
+            if (perc) begin_f = true;
+        }
+        if (num_params != count) throw "Number of format specifiers and arguments must match";
```
We should be able to simplify this.

-----

Suggestion 1: we swap the state machine when encountering a `%`, otherwise inc the param count and step out of the format:
```C++
constexpr static void CheckNumFormatSpecifiers(std::string_view str)
{
    int count{0};
    bool in_format{false};
    for (char c : str) {
        if (c == '%') in_format = !in_format;
        else if (in_format) {
            ++count;
            in_format = false;
        }
    }
    if (num_params != count) throw std::runtime_error("Format specifier count must match the argument count!");
 }
```

----

Suggestion 2: count the excesses and throw when there is any:
```C++
constexpr static void CheckNumFormatSpecifiers(std::string_view str)
    int excess{num_params};
    bool in_format{false};
    for (char c : str) {
        if (c == '%') in_format = !in_format;
        else if (in_format) {
            --excess;
            in_format = false;
        }
    }
    if (excess) throw std::runtime_error("Format specifier count must match the argument count!");
 }
```

l0rinc replied:
By counting the excess we could give more fine-grained errors, e.g:
```C++
constexpr static void Detail_CheckNumFormatSpecifiers(std::string_view str)
{
    int excess{num_params};
    bool in_format{false};
    for (char c: str) {
        if (c == '%') in_format = !in_format;
        else if (in_format) {
            --excess;
            in_format = false;
        }
    }
    if (in_format) throw "Format specifier incorrectly terminated by end of string!";
    if (excess > 0) throw "Too few format specifiers!";
    if (excess < 0) throw "Too many format specifiers!";
}
```

and tests like:
```C++
auto check_too_many_specifiers = [](ErrType str) { return str == "Too many format specifiers!"s; };
BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%_"), ErrType, check_too_many_specifiers);

auto check_too_few_specifiers = [](ErrType str) { return str == "Too few format specifiers!"s; };
BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers(""), ErrType, check_too_few_specifiers);
BOOST_CHECK_EXCEPTION(ConstevalFormatString<2>::Detail_CheckNumFormatSpecifiers("%_"), ErrType, check_too_few_specifiers);
```

l0rinc replied:
> important thing is the error, not the exact wording of the error message

You have separate error for trailing `%`, why not have dedicated errors for when the format specifiers are fewer/more than the args?

l0rinc replied:
I saw your previous response, I was quoting from it in mine. Anyway, you can resolve this comment.

l0rinc commented on src/util/string.h:
```diff
@@ -17,6 +17,28 @@
+    consteval ConstevalFormatString(const char* str) : fmt{str} { CheckNumFormatSpecifiers(fmt); }
+    consteval static void CheckNumFormatSpecifiers(std::string_view str)
+    {
+        int count{0};
+        bool begin_f{false};
+        for (char c : str) {
+            bool perc{c == '%'};
+            if (begin_f) {
+                count += !perc;
+                begin_f = false;
+                continue;
+            }
+            if (perc) begin_f = true;
+        }
+        if (num_params != count) throw "Number of format specifiers and arguments must match";
```
if we throw a `runtime_error`, testing becomes slightly simpler:
```C++
std::runtime_error("Format specifier count must match the argument count!")
```
otherwise it's:
```C++
BOOST_CHECK_THROW(ConstevalFormatString<1>::CheckNumFormatSpecifiers(""), const char*);
```

nit: slight rewording

l0rinc replied:
How come? Wouldn't that be a simpler way to check that we're throwing in this case (instead of the constant err and err_types which cannot even change)
```C++
BOOST_CHECK_THROW(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers(""), const char*);
```

l0rinc replied:
Seems a bit excessive to me, could we maybe specialize the tests when those errors are actually added?

l0rinc reviewed (changes requested):
Approach NACK, it seems to me there's a way to properly test this (including failure cases):

<details>
<summary>Details</summary>

```patch
Index: src/test/string_tests.cpp
<+>UTF-8
===================================================================
diff --git a/src/test/string_tests.cpp b/src/test/string_tests.cpp
new file mode 100644
--- /dev/null(date 1725028151058)
+++ b/src/test/string_tests.cpp(date 1725028151058)
@@ -0,0 +1,52 @@
+#include <boost/test/unit_test.hpp>
+#include <index/base.h>
+#include <test/util/setup_common.h>
+#include <univalue.h>
+#include <util/string.h>
+
+#include <string>
+
+using namespace util;
+
+BOOST_FIXTURE_TEST_SUITE(string_tests, BasicTestingSetup)
+
+BOOST_AUTO_TEST_CASE(CheckNumFormatSpecifiers_test)
+{
+    // No format specifiers
+    ConstevalFormatString<0>::CheckNumFormatSpecifiers("");
+    ConstevalFormatString<0>::CheckNumFormatSpecifiers("%");
+    ConstevalFormatString<0>::CheckNumFormatSpecifiers("%%");
+    ConstevalFormatString<0>::CheckNumFormatSpecifiers("%%%");
+    ConstevalFormatString<0>::CheckNumFormatSpecifiers("_%");
+    ConstevalFormatString<0>::CheckNumFormatSpecifiers("_%%");
+    ConstevalFormatString<0>::CheckNumFormatSpecifiers("%%_%");
+    ConstevalFormatString<0>::CheckNumFormatSpecifiers("_%%%");
+
+    // Single format specifier
+    ConstevalFormatString<1>::CheckNumFormatSpecifiers("%d");
+    ConstevalFormatString<1>::CheckNumFormatSpecifiers("%_");
+    ConstevalFormatString<1>::CheckNumFormatSpecifiers("%_%");
+    ConstevalFormatString<1>::CheckNumFormatSpecifiers("%%%_");
+    ConstevalFormatString<1>::CheckNumFormatSpecifiers("%_%%");
+    ConstevalFormatString<1>::CheckNumFormatSpecifiers("%d%%");
+
+    // Multiple format specifiers
+    ConstevalFormatString<2>::CheckNumFormatSpecifiers("%d%s");
+    ConstevalFormatString<3>::CheckNumFormatSpecifiers("%d%s%f");
+    ConstevalFormatString<2>::CheckNumFormatSpecifiers("%%d%d%i");
+    ConstevalFormatString<3>::CheckNumFormatSpecifiers("%d%s%%d%i");
+
+    // Positional arguments
+    ConstevalFormatString<2>::CheckNumFormatSpecifiers("%1$d%2$s");
+    ConstevalFormatString<3>::CheckNumFormatSpecifiers("%3$d%1$s%2$f");
+    ConstevalFormatString<3>::CheckNumFormatSpecifiers("%3$d%%d%1$s%2$f");
+    ConstevalFormatString<2>::CheckNumFormatSpecifiers("%1$d%s");
+
+    // Negative cases
+    BOOST_CHECK_THROW(ConstevalFormatString<1>::CheckNumFormatSpecifiers(""), std::runtime_error);
+    BOOST_CHECK_THROW(ConstevalFormatString<0>::CheckNumFormatSpecifiers("%d"), std::runtime_error);
+    BOOST_CHECK_THROW(ConstevalFormatString<2>::CheckNumFormatSpecifiers("%d"), std::runtime_error);
+    BOOST_CHECK_THROW(ConstevalFormatString<3>::CheckNumFormatSpecifiers("%1$d%2$s"), std::runtime_error);
+}
+
+BOOST_AUTO_TEST_SUITE_END()
\ No newline at end of file
Index: src/util/string.h
<+>UTF-8
===================================================================
diff --git a/src/util/string.h b/src/util/string.h
--- a/src/util/string.h(revision 60c75975b55adcce015a857c529baafd65b58394)
+++ b/src/util/string.h(date 1725027928051)
@@ -22,20 +22,18 @@
 struct ConstevalFormatString {
     const char* const fmt;
     consteval ConstevalFormatString(const char* str) : fmt{str} { CheckNumFormatSpecifiers(fmt); }
-    consteval static void CheckNumFormatSpecifiers(std::string_view str)
+    constexpr static void CheckNumFormatSpecifiers(std::string_view str)
     {
-        int count{0};
-        bool begin_f{false};
+        int excess{num_params};
+        bool in_format{false};
         for (char c : str) {
-            bool perc{c == '%'};
-            if (begin_f) {
-                count += !perc;
-                begin_f = false;
-                continue;
+            if (c == '%') in_format = !in_format;
+            else if (in_format) {
+                --excess;
+                in_format = false;
             }
-            if (perc) begin_f = true;
         }
-        if (num_params != count) throw "Number of format specifiers and arguments must match";
+        if (excess) throw std::runtime_error("Format specifier count must match the argument count!");
     }
 };
 
Index: src/test/CMakeLists.txt
<+>UTF-8
===================================================================
diff --git a/src/test/CMakeLists.txt b/src/test/CMakeLists.txt
--- a/src/test/CMakeLists.txt(revision 60c75975b55adcce015a857c529baafd65b58394)
+++ b/src/test/CMakeLists.txt(date 1725025275931)
@@ -119,6 +119,7 @@
   sock_tests.cpp
   span_tests.cpp
   streams_tests.cpp
+  string_tests.cpp
   sync_tests.cpp
   system_tests.cpp
   timeoffsets_tests.cpp
Index: src/util/string.cpp
<+>UTF-8
===================================================================
diff --git a/src/util/string.cpp b/src/util/string.cpp
--- a/src/util/string.cpp(revision 60c75975b55adcce015a857c529baafd65b58394)
+++ b/src/util/string.cpp(date 1725025405333)
@@ -9,24 +9,6 @@
 
 namespace util {
 
-// Compile-time sanity checks
-static_assert([] {
-    ConstevalFormatString<0>::CheckNumFormatSpecifiers("");
-    ConstevalFormatString<0>::CheckNumFormatSpecifiers("%");
-    ConstevalFormatString<0>::CheckNumFormatSpecifiers("%%");
-    ConstevalFormatString<0>::CheckNumFormatSpecifiers("%%%");
-    ConstevalFormatString<1>::CheckNumFormatSpecifiers("%_");
-    ConstevalFormatString<0>::CheckNumFormatSpecifiers("_%");
-    ConstevalFormatString<0>::CheckNumFormatSpecifiers("%%_");
-    ConstevalFormatString<1>::CheckNumFormatSpecifiers("%_%");
-    ConstevalFormatString<0>::CheckNumFormatSpecifiers("_%%");
-    ConstevalFormatString<1>::CheckNumFormatSpecifiers("%%%_");
-    ConstevalFormatString<0>::CheckNumFormatSpecifiers("%%_%");
-    ConstevalFormatString<1>::CheckNumFormatSpecifiers("%_%%");
-    ConstevalFormatString<0>::CheckNumFormatSpecifiers("_%%%");
-    return true; // All checks above compiled and passed
-}());
-
 void ReplaceAll(std::string& in_out, const std::string& search, const std::string& substitute)
 {
     if (search.empty()) return;
```

</details>

l0rinc commented on src/util/string.cpp:
```diff
@@ -8,6 +8,25 @@
+    ConstevalFormatString<0>::CheckNumFormatSpecifiers("");
+    ConstevalFormatString<0>::CheckNumFormatSpecifiers("%");
+    ConstevalFormatString<0>::CheckNumFormatSpecifiers("%%");
+    ConstevalFormatString<0>::CheckNumFormatSpecifiers("%%%");
```
shouldn't we warn in this case as well, i.e. an unfinished %?
The tinyformat tests use two trailing % when it's an actual %:
https://github.com/c42f/tinyformat/blob/master/tinyformat_test.cpp#L144

hodlinator commented on src/util/string.h:
```diff
@@ -17,6 +17,28 @@
+struct ConstevalFormatString {
+    const char* const fmt;
+    consteval ConstevalFormatString(const char* str) : fmt{str} { CheckNumFormatSpecifiers(fmt); }
+    consteval static void CheckNumFormatSpecifiers(std::string_view str)
```
One could potentially keep it `consteval` and have `CheckNumFormatSpecifiers` return a `bool` to provide for negatives, but the error messaging would suffer.

hodlinator reviewed (changes requested):
Concept ACK faafbcd5e7245586e3f002147b65cc6133bf9a6f

#### Specifier validation

**run-lint-format-strings.py** does handle more complex format specifications through the ["%(.*?)[aAcdeEfFgGinopsuxX]" regexp](https://github.com/bitcoin/bitcoin/pull/30546/files#diff-bdbca6eab11048c0ffaa903b5ee733cec028f10051bf92cf9da966412669e9c9L253). (https://en.cppreference.com/w/cpp/io/c/fprintf for details on what should be supported).

Unsupported conversion specifiers are printed as-is by `printf`:
`printf("%1 %H %d %K %k", 10); -> "%1 %H 10 %K %k"`
Should easily be made into errors:
`std::string_view{"aAcdeEfFgGinopsuxX"}.find(c) == std::string_view::npos`

(Having something between the %-sign and a valid conversion specifier-letter should probably be an error if we don't handle it).

#### (Type safety)

It would be really cool if we also checked for type safety à la [__attribute__ ((format (printf, ...)))](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-format-function-attribute) but **run-lint-format-strings.py** primarily attempts to verify that the count of args matched. Better done as a follow-up PR if someone's up for it.

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -0,0 +1,44 @@
+
+BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
+{
+    // Compile-time sanity checks
+    static_assert([] {
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%%");
+        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("_%");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%_");
+        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_%");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("_%%");
+        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%%%_");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%_%");
+        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_%%");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("_%%%");
+        return true; // All checks above compiled and passed
+    }());
```
the errors are really ugly this way:
> util_string_tests.cpp:16:19: error: static assertion expression is not an integral constant expression
    static_assert([] {
                  ^~~~

Would there be any disadvantage in inlining them?
```C++
BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
{
    ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("");
    ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%");
...
    ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_%%");
    ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("_%%%");
```

l0rinc replied:
While the code is slightly simpler (-4 lines, indentation), the error message in case of failure is also simpler:

<details>
<summary>with static assert</summary>

```
bitcoin/src/test/util_string_tests.cpp:16:19: error: static assertion expression is not an integral constant expression
    static_assert([] {
                  ^~~~
bitcoin/src/util/string.h:39:34: note: subexpression not valid in a constant expression
        if (num_params != count) throw "Format specifier count must match the argument count!";
                                 ^
bitcoin/src/test/util_string_tests.cpp:22:9: note: in call to 'Detail_CheckNumFormatSpecifiers({&"% % %_"[0], 6})'
        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("% % %_");
        ^
bitcoin/src/test/util_string_tests.cpp:16:19: note: in call to '&[] {
    ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("");
    ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%");
    ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_");
    ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%_");
    ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("_%%");
    ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("% % %_");
    ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_%%");
    return true;
}->operator()()'
    static_assert([] {
                  ^
1 error generated.
make[2]: *** [src/test/CMakeFiles/test_bitcoin.dir/util_string_tests.cpp.o] Error 1
make[1]: *** [src/test/CMakeFiles/test_bitcoin.dir/all] Error 2
make: *** [all] Error 2
```

</details>

<details>
<summary>without static assert</summary>

```
unknown location:0: fatal error: in "util_string_tests/ConstevalFormatString_NumSpec": C string: Format specifier count must match the argument count!
src/test/util_string_tests.cpp:13: last checkpoint: "ConstevalFormatString_NumSpec" test entry

*** 1 failure is detected in the test module "Bitcoin Core Test Suite"
```

</details>

> Moving the checks from being done at compile-time to be done at runtime means that possible errors will be caught later

In reality these lines will rarely be modified - but probably read more often, so I'd go for the simpler code.
If you still insist that this is better, please resolve the comment.

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -0,0 +1,44 @@
+    }());
+
+    // Negative checks at runtime
+    constexpr auto err{"Number of format specifiers and arguments must match"};
+    using err_type = decltype(err);
+    struct CheckThrow {
+        bool operator()(const err_type& str) const { return std::string_view{str} == err; }
+    };
+    BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers(""), err_type, CheckThrow{});
+    BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%_"), err_type, CheckThrow{});
+    BOOST_CHECK_EXCEPTION(ConstevalFormatString<2>::Detail_CheckNumFormatSpecifiers("%_"), err_type, CheckThrow{});
```
As mentioned in another comment, wouldn't this be simpler:
```C++
    BOOST_CHECK_THROW(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers(""), const char*);
    BOOST_CHECK_THROW(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%_"), const char*);
    BOOST_CHECK_THROW(ConstevalFormatString<2>::Detail_CheckNumFormatSpecifiers("%_"), const char*);
```
?

l0rinc commented on src/util/string.cpp:
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2019-2022 The Bitcoin Core developers
+// Copyright (c) 2019-present The Bitcoin Core developers
```
is this a leftover?

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -0,0 +1,48 @@
+    // Negative checks at runtime
+    using ErrType = const char*;
+
+    struct CheckThrowNumSpec {
+        bool operator()(const ErrType& str) const { return std::string_view{str} == "Format specifier count must match the argument count!"; }
+    };
```
Instead of a functor I think we can simplify this by using a lambda instead.
And we can also simplify the comparison by using string_literals, i.e.:
```C++
#include <string>
using namespace std::string_literals;
...
    auto check_throw_num_spec = [](ErrType str) { return str == "Format specifier count must match the argument count!"s; };
    BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers(""), ErrType, check_throw_num_spec);
...
```

l0rinc replied:
What's the advantage of the `string_view` version?

l0rinc replied:
ok, that's fair, thanks.

l0rinc reviewed (commented):
Concept ACK, the remaining nits are non-blocking

l0rinc commented on src/util/string.h:
```diff
@@ -17,6 +17,29 @@
 
 namespace util {
+/** Type to denote a format string that was checked at compile time */
+template <int num_params>
+struct ConstevalFormatString {
+    const char* const fmt;
+    consteval ConstevalFormatString(const char* str) : fmt{str} { Detail_CheckNumFormatSpecifiers(fmt); }
+    constexpr static void Detail_CheckNumFormatSpecifiers(std::string_view str)
+    {
+        int count{0};
```
If you keep this version, `num_params` can likely be a `size_t` instead:
```diff
diff --git a/src/util/string.h b/src/util/string.h
--- a/src/util/string.h	(revision fa72ce66421d3f90a6794b3e54e56873ae81265f)
+++ b/src/util/string.h	(date 1725561949080)
@@ -18,13 +18,13 @@
 
 namespace util {
 /** Type to denote a format string that was checked at compile time */
-template <int num_params>
+template <size_t num_params>
 struct ConstevalFormatString {
     const char* const fmt;
     consteval ConstevalFormatString(const char* str) : fmt{str} { Detail_CheckNumFormatSpecifiers(fmt); }
     constexpr static void Detail_CheckNumFormatSpecifiers(std::string_view str)
     {
-        int count{0};
+        size_t count{0};
         bool begin_f{false};
         for (char c : str) {
             bool perc{c == '%'};
```

l0rinc replied:
it's not the counting, it's to document that this version doesn't have negatives (like the excess version does)

hodlinator commented:
#### Invalid chars check

Here follows an implementation of the low bar of what I stated in https://github.com/bitcoin/bitcoin/pull/30546#pullrequestreview-2273568357. It's implemented on top of fa72ce66421d3f90a6794b3e54e56873ae81265f.

<details>
<summary>

##### Diff

</summary>

I've confirmed that "aAcdfeEfFgGinopsuxX" matches what I linked to in cppreference and also the Python version being replaced.

```diff

diff --git a/src/test/util_string_tests.cpp b/src/test/util_string_tests.cpp
index ecc28f5b96..9254f57b64 100644
--- a/src/test/util_string_tests.cpp
+++ b/src/test/util_string_tests.cpp
@@ -16,11 +16,11 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
     static_assert([] {
         ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("");
         ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%");
-        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_");
-        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%_");
-        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("_%%");
-        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%%%_");
-        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_%%");
+        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%s");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%s");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("s%%");
+        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%%%s");
+        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%s%%");
         return true; // All checks above compiled and passed
     }());
 
@@ -31,18 +31,25 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
         bool operator()(const ErrType& str) const { return std::string_view{str} == "Format specifier count must match the argument count!"; }
     };
     BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers(""), ErrType, CheckThrowNumSpec{});
-    BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%_"), ErrType, CheckThrowNumSpec{});
-    BOOST_CHECK_EXCEPTION(ConstevalFormatString<2>::Detail_CheckNumFormatSpecifiers("%_"), ErrType, CheckThrowNumSpec{});
+    BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%s"), ErrType, CheckThrowNumSpec{});
+    BOOST_CHECK_EXCEPTION(ConstevalFormatString<2>::Detail_CheckNumFormatSpecifiers("%s"), ErrType, CheckThrowNumSpec{});
 
     struct CheckThrowTerm {
         bool operator()(const ErrType& str) const { return std::string_view{str} == "Format specifier incorrectly terminated by end of string!"; }
     };
     BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%"), ErrType, CheckThrowTerm{});
     BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%%"), ErrType, CheckThrowTerm{});
-    BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("_%"), ErrType, CheckThrowTerm{});
-    BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_%"), ErrType, CheckThrowTerm{});
-    BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%_%"), ErrType, CheckThrowTerm{});
-    BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("_%%%"), ErrType, CheckThrowTerm{});
+    BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("s%"), ErrType, CheckThrowTerm{});
+    BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%s%"), ErrType, CheckThrowTerm{});
+    BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%s%"), ErrType, CheckThrowTerm{});
+    BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("s%%%"), ErrType, CheckThrowTerm{});
+
+    struct CheckThrowInvalid {
+        bool operator()(const ErrType& str) const { return std::string_view{str} == "Invalid format specifier!"; }
+    };
+    BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%2"), ErrType, CheckThrowInvalid{});
+    BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%$"), ErrType, CheckThrowInvalid{});
+    BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%z"), ErrType, CheckThrowInvalid{});
 }
 
 BOOST_AUTO_TEST_SUITE_END()
diff --git a/src/util/string.h b/src/util/string.h
index e968de5575..5ae54e1217 100644
--- a/src/util/string.h
+++ b/src/util/string.h
@@ -29,6 +29,9 @@ struct ConstevalFormatString {
         for (char c : str) {
             bool perc{c == '%'};
             if (begin_f) {
+                if (!perc && std::string_view{"aAcdfeEfFgGinopsuxX"}.find(c) == std::string_view::npos) {
+                    throw "Invalid format specifier!";
+                }
                 count += !perc;
                 begin_f = false;
                 continue;
```
</details>

l0rinc reviewed (approved):
While I would still prefer some of our comments to be reconsidered, I’m also fine with merging the change as is.
Thanks for streamlining the developer experience with these small but useful checks!

ACK fa092749094aa483e3ce0243885ce2eb8ed22cbb

hodlinator commented:
> In tinyformat there is no such thing, so I'll leave this as-is for now.

My motivation is not to reach parity with *tinyformat*, it is to at least cover some of what the prior linter did:

https://github.com/bitcoin/bitcoin/blob/c3af4b1ec3fdb308404199ddd0df5170793a2c39/test/lint/run-lint-format-strings.py#L230-L253

An alternative to making it an error is to simply make invalid chars not count, which is closer to the old behavior. Would that be okay?

If you don't want to bring over at least some of that specific functionality, the PR summary should state why.

l0rinc reviewed (approved):
ACK fa18f7cc9ac36ebe0fa8c381b1db76085812e95f

l0rinc commented on src/util/string.h:
```diff
@@ -17,6 +17,54 @@
+                    count_nor += !perc;
+                } else {
+                    // Positional specifier
+                    if (*it_num != '$' || std::string_view{"aAcdfeEfFgGinopsuxX"}.find(*(++it_num)) == std::string_view::npos) {
```
Agree, in C++23 we can just use `.contains` which would simplify this slightly

stickies-v reviewed (commented):
Concept ACK. Doing this at compile time is much more helpful, besides the bugs being fixed.

l0rinc commented on src/util/string.h:
```diff
@@ -173,4 +222,12 @@ template <typename T1, size_t PREFIX_LEN>
 
+namespace tinyformat {
+template <typename... Args>
+std::string format(util::ConstevalFormatString<sizeof...(Args)> fmt, const Args&... args)
```
can we still have errors like [this](https://github.com/bitcoin/bitcoin/blob/master/src/logging.h#L253-L265) after this change?

l0rinc commented:
ACK faa32adbcf4c04f0a426eaba4a43b29a293de72b

<details>
<summary>Details</summary>

* brace initialization in base.cpp
* pos*t*itional typo fix
* added ConstevalFormatString  overload

</details>

hodlinator commented on src/test/util_string_tests.cpp:
```diff
@@ -0,0 +1,57 @@
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("s%%");
+        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%%%s");
+        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%s%%");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers(" 1$s");
```
Tried adding
```
        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%02d");
```
here and it fails.

Ran
`git grep -E '"[^"]*%[0-9]+[aAcdfeEfFgGinopsuxX*][^"]*' src/ ':(exclude)src/leveldb' ':(exclude)src/secp256k1' ':(exclude)*.ts'`
and it feels like a common enough case to want to have fixed width formatting (*optionally* `0`-padded). Examples:
```
src/netbase.cpp:            LogError("Proxy requested wrong authentication method %02x\n", pchRet1[1]);
src/flatfile.cpp:                LogDebug(BCLog::VALIDATION, "Pre-allocating up to position 0x%x in %s%05u.dat\n", new_size, m_prefix, pos.nFile);
src/init.cpp:                LogPrintf(" block index %15dms\n", Ticks<std::chrono::milliseconds>(SteadyClock::now() - load_block_index_start_time));
src/node/blockstorage.cpp:            LogDebug(BCLog::BLOCKSTORAGE, "Prune: %s deleted blk/rev (%05u)\n", __func__, *it);
src/node/blockstorage.cpp:                          "Failed to flush previous block file %05i (finalize=1, finalize_undo=%i) before opening new block file %05i\n",
src/node/blockstorage.cpp:                LogPrintLevel(BCLog::BLOCKSTORAGE, BCLog::Level::Warning, "Failed to flush undo file %05i\n", _pos.nFile);
src/node/blockstorage.cpp:            LogPrintf("Reindexing block file blk%05u.dat...\n", (unsigned int)nFile);
```

Sorry for not realizing and raising this in an earlier review.

hodlinator reviewed (commented):
Reviewed faa32adbcf4c04f0a426eaba4a43b29a293de72b

`git range-diff master fa72ce6 faa32ad`

- Implemented positional args after me finally communicating clearly enough. 
- Added `ConstevalFormatString` overload directly into the `tinyformat` namespace to ease future adoption.
- Uses more terse lambdas instead of explicit `struct`s for checking exceptions. :+1: 

Does not implement support for '*', but not currently used anyway:
https://github.com/bitcoin/bitcoin/blob/c3af4b1ec3fdb308404199ddd0df5170793a2c39/test/lint/run-lint-format-strings.py#L247

New concern: Does not support supplying an integer to specify width together with an optional `0` to change padding from spaces to zeroes. (See inline comment).

stickies-v commented on src/util/string.h:
```diff
@@ -17,6 +18,54 @@
+                // terminator in the format string "%\n". Also, excluding '0'
+                // avoids an out-of-range position error in tinyformat for the
+                // format string "%0$s\n".
+                if (std::string_view{"123456789aAcdfeEfFgGinopsuxX%"}.find(*it) == std::string_view::npos) throw "Invalid format specifier!";
```
We do have quite a few use cases of `\.[\d+]f` precision that would throw here. No issue in the current code, but would like have to be overhauled in the next PR?

stickies-v reviewed (commented):
I'm still a strong Concept ACK on this, but I wonder if the incremental appoach currently taken is the best approach. We'll have to (I think quite likely non-trivially) update the logic based on flags, widths, precisions, conversion, ... actually used in the codebase, as seen with e.g. the [precision](https://github.com/bitcoin/bitcoin/pull/30546#discussion_r1752314431) and [zero-padding](https://github.com/bitcoin/bitcoin/pull/30546#discussion_r1752026952) comments?

So perhaps it'd be sensible to first create a draft PR/branch that forces `ConstevalFormatString` everywhere, informing our requirements and adding full unit test coverage, and then once we have a good overview, branch off the minimal subset of changes required to update `FatalErrorf` as to keep that first (i.e. this) PR merge-able?

Commit:
lint: Remove forbidden functions from lint-format-strings.py

Given that all of them are forbidden by the
test/lint/lint-locale-dependence.py check, they can be removed.

stickies-v commented on src/util/string.h:
```diff
@@ -17,6 +17,54 @@
+        }
+        if (count_nor && count_pos) throw "Format specifiers must be all postitional or all non-positional!";
+        unsigned count{count_nor | count_pos};
+        if (num_params != count) throw "Format specifier count must match the argument count!";
```
> making implementation details verbose to use is a feature to avoid developers from accidentally using them in real code

I can't immediately see a use case for this function, but is there a reason we'd want to actively discourage this function from being used?  It seems harmless to me, but if there is, perhaps it'd be good to add a brief docstring to the function to describe that?

> The verbosity in tests is a bit annoying, but seems acceptable to me.

I agree.

stickies-v commented on src/util/string.h:
```diff
@@ -17,6 +17,54 @@
+        unsigned count_pos{0}; // Max number in positional specifier, like %8$s
+        bool begin_f{false};
+        for (auto it{str.begin()}; it < str.end(); ++it) {
+            bool perc{*it == '%'};
```
> Given that they are not used,

Yup, I agree with the simplified approach. I didn't like the half/half of quite a of bit of complexity without being complete, this makes more sense.  Can be marked as resolved, thanks!

Commit:
util: Use compile-time check for LogConnectFailure

stickies-v commented on src/util/string.h:
```diff
@@ -17,6 +18,55 @@
 #include <vector>
 
 namespace util {
+/** Type to denote a format string that was checked at compile time */
```
I think this docstring could be beefed up. To give it a start:

```
/**
 * @brief A wrapper for a compile-time partially validated format string
 *
 * This struct can be used to enforce partial compile-time validation of format strings, to reduce
 * the likelihood of tinyformat throwing exceptions at run-time. Validation is partial to try and
 * prevent the most common errors while avoiding re-implementing the entire parsing logic.
 * 
 * @note Counting of `*` dynamic width and precision fields (such as `%*c`, `%2$*3$d`, `%.*f`) is
 * notably not implemented to minimize code complexity as long they are not used in the codebase.
 * Usage of these fields is not counted and can lead to run-time exceptions.
 */
```

stickies-v reviewed (commented):
Approach ACK and code LGTM faca9a821963d629987f6c2a2f7d13b1bf01162c , left some suggestions that would be nice.

PR description also needs a bit of updating for the 2 new commits re forbidden functions and covering `LogConnectFailure`. The 2 new commits don't seem necessary for this PR but they were trivial to review and relevant to the goal so I'm okay keeping them.

l0rinc commented on src/util/string.h:
```diff
@@ -17,6 +18,55 @@
+struct ConstevalFormatString {
+    const char* const fmt;
+    consteval ConstevalFormatString(const char* str) : fmt{str} { Detail_CheckNumFormatSpecifiers(fmt); }
+    constexpr static void Detail_CheckNumFormatSpecifiers(std::string_view str)
```
While it gets the job done, I think the parser is unnecessarily complicated for several reasons:
* Redundant Counters: Separate (but mutually exclusive) `count_normal` and `count_pos`.
* Repetition: The check for "Format specifier incorrectly terminated by end of string" is repeated in multiple places.
* Inefficient Character Handling: Multiple character checks for `%` (and manual digit detection).
* Scattered Iterations: Increment operations (++it), continue statements, and lookaheads (str.end()) make the control flow harder to follow.

I have a hard time explaining the behavior of the parser in words, following the logic of the code. Can we use a state-machine based parser instead, without lookaround, iterating char by char without jumping around, changing the state as we go along, e.g. something like this:
```C++
constexpr static void Detail_CheckNumFormatSpecifiers(std::string_view str)
{
    unsigned count{0};
    unsigned maybe_num{0};
    bool inside_format_specifier = false;
    bool has_normal = false, has_positional = false;
    for (auto it = str.begin(); it < str.end(); ++it) {
        if (*it == '%') inside_format_specifier = !inside_format_specifier;
        else if (inside_format_specifier) {
            if (std::isdigit(*it)) maybe_num = maybe_num * 10 + (*it - '0');
            else if (*it == '$') {
                if (maybe_num == 0) throw "Positional format specifier must be at least 1";
                count = std::max(count, maybe_num);
                has_positional = true;
                inside_format_specifier = false;
            } else {
                ++count;
                maybe_num = 0;
                has_normal = true;
                inside_format_specifier = false;
            }
        } else maybe_num = 0;
    }

    if (has_normal && has_positional) throw "Format specifiers must be all positional or all non-positional!";
    if (inside_format_specifier || maybe_num > 0) throw "Format specifier incorrectly terminated by end of string";
    if (num_params != count) throw "Format specifier count must match the argument count!";
}
```

This is the reason I'm not ack-ig yet.

l0rinc replied:
Thanks, updated the example

l0rinc commented on src/util/string.h:
```diff
@@ -17,6 +18,55 @@
+            if (*it_num == '$') {
+                // Positional specifier
+                if (maybe_num == 0) throw "Positional format specifier must indicate at least 1";
+                count_pos = std::max(count_pos, maybe_num);
```
Alternatively, we could cheaply collect the referenced indexes to a `std::set<unsigned> referenced_params;` and check that all of them were referenced, i.e. `if (referenced_params.size() != num_params) throw "Not all parameters are referenced!";`

l0rinc commented on src/util/string.h:
```diff
@@ -17,6 +18,55 @@
+                continue;
+            }
+
+            if (++it >= str.end()) throw "Format specifier incorrectly terminated by end of string";
```
I don't particularly like that we have lookaheads even though we should be able to detect this at the very end - which would enable us to remove the duplicated throw.

l0rinc commented on src/util/string.h:
```diff
@@ -17,6 +18,55 @@
+        unsigned count_pos{0};    // Max number in positional specifier, like %8$s
+        for (auto it{str.begin()}; it < str.end();) {
+            if (*it != '%') {
+                ++it;
+                continue;
```
I find these jumps quite confusing, it doesn't reflect how I would manually read and interpret such a format string

l0rinc reviewed (changes requested):
Thanks for your patience and for striving to eliminate developer frustration.
I like the concept a lot, but have some preferences regarding parser intuitiveness, hope you'll consider my suggestions.

l0rinc commented on src/util/string.h:
```diff
@@ -17,6 +18,55 @@
+            }
+        }
+        if (count_normal && count_pos) throw "Format specifiers must be all positional or all non-positional!";
+        unsigned count{count_normal | count_pos};
```
The error messages indicate two separate concerns, it would make sense logically to separate them as well

l0rinc commented on src/util/string.h:
```diff
@@ -17,6 +17,55 @@
+                continue;
+            }
+
+            // Test for positional specifier, like %8$
```
Absolutely, but we may not want to spend time implementing positional parsing, if we agree that we don't want to encourage their usage

ryanofsky replied:
I think positional args are more readable than repeated args (even in this example which is atypically long) because:

- People have limited working memories. it is a easier to see 4 parameters passed and remember what numbers 1-4 are than to see 6 parameters passed and remember what 1-6 are. Remembering a list is also harder if the list has duplicate values.
- Positional arguments are more explicit. "Configuration file %3$s" it is clearer than "Configuration file %s" because it tells you that the 3rd argument is substituted. You don't have to manually scan the string and count earlier %s.
- Positional arguments make it easier understand messages and see whether they are internally consistent. "Data directory %1$s contains a %2$s file [...] Delete or rename the %2$s file in data directory %1$s" is easier to understand and verify than the alternative with %s everywhere.

l0rinc replied:
Thanks @ryanofsky, it seems there is a need for that after all, please resolve this comment.

hodlinator reviewed (commented):
Reviewed fa7819bfadd29e41a1c9283828f7e32934d4fbd9

`git range-diff master faa32ad fa7819b`

- Removed validation of `aAcdfeEfFgGinopsuxX` as tinyformat does not implement it either - acceptable (Edit: parity with tinyformat is more important than parity with a custom linter for tinyformat).
- Implemented support for width/padding-specification.
- Now applies to `LogConnectFailure`.
- Adds commit fafe6313cce2b97d86bd6f907ffe93c98cf9cb59 which deduplicates lints.

#### fafe6313cce2b97d86bd6f907ffe93c98cf9cb59 - lint: Remove forbidden functions from lint-format-strings.py

Could move this commit first since it is not dependent on the prior commits?

ryanofsky reviewed (approved):
Code review ACK faa5a8e4480de69086796dba49b65ef73058f8d7

I like this approach of implementing a loose compile time checker that avoids too much complexity by supporting a superset of tinyformat syntax. This approach is more flexible and should be easier to roll out than the opposite approach which would avoid complexity by only supporting a subset of tinyformat syntax.

In longer run though, I think just supporting a subset of tinyformat syntax would be preferable to provide more safety, and make the checking code easier to understand. @l0rinc seems to have some good ideas about this and be pretty eager to rewrite the checking function, so there should be opportunities to make future improvements after this is used more widely.

ryanofsky reviewed (approved):
Code review ACK fa26462e95291652b4021d91b014655f678149e8. Just new tests and comments since last review

hodlinator reviewed (commented):
Reviewed fa26462e95291652b4021d91b014655f678149e8

`git range-diff master fa7819b fa26462`

- Moved independent commit to the beginning, thanks!
- Added some comments.

I really like that this is being worked upon and want to help us find the best trade-offs in settling on a solution.

hodlinator commented on src/test/util_string_tests.cpp:
```diff
@@ -0,0 +1,80 @@
+
+BOOST_AUTO_TEST_SUITE(util_string_tests)
+
+BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
```
> I think that real code should be designed without taking tests into account (apart from being testable). Tests should always follow real code, not the other way round.

Making the function return a value allowed for C) dropped negatives (and less runtime testing). But I can try to provide a new diff bringing them back and without changing *string.h*. My main point remains adding the interlaced tinyformat tests to prove parity.

> The diff you shared also changes `util::hex_literals::detail::Hex`, which I don't really understand.

It's a consequence of clashes between `util::detail` namespaces. (Added note about that in a later edit).

> All relevant test coverage has been taken over, except for the `*` test coverage. But the `*` parsing is broken in the linter anyway, so I am not sure about the value of taking over test in a comment (dead code).

Fair.

hodlinator replied:
<details>
<summary>New <code>git diff -w</code></summary>

```diff
diff --git a/src/test/util_string_tests.cpp b/src/test/util_string_tests.cpp
index c6ee1ffed9..3fed361e7e 100644
--- a/src/test/util_string_tests.cpp
+++ b/src/test/util_string_tests.cpp
@@ -10,62 +10,140 @@ using namespace util;
 
 BOOST_AUTO_TEST_SUITE(util_string_tests)
 
-BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
+// These are counted correctly, since tinyformat will require the provided number of args.
+BOOST_AUTO_TEST_CASE(ConstevalFormatString_CorrectCounts)
 {
-    // Compile-time sanity checks
-    static_assert([] {
     ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("");
+    tfm::format("");
+
     ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%");
+    tfm::format("%%");
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%s");
+    BOOST_CHECK_THROW(tfm::format("%s"), tfm::format_error);
+    tfm::format("%s", "foo");
+
     ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%s");
+    tfm::format("%%s");
+
     ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("s%%");
+    tfm::format("s%%");
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%%%s");
+    BOOST_CHECK_THROW(tfm::format("%%%s"), tfm::format_error);
+    tfm::format("%%%s", "foo");
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%s%%");
+    BOOST_CHECK_THROW(tfm::format("%s%%"), tfm::format_error);
+    tfm::format("%s%%", "foo");
+
     ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers(" 1$s");
+    tfm::format(" 1$s");
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%1$s");
+    BOOST_CHECK_THROW(tfm::format("%1$s"), tfm::format_error);
+    tfm::format("%1$s", "foo");
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%1$s%1$s");
+    BOOST_CHECK_THROW(tfm::format("%1$s%1$s"), tfm::format_error);
+    tfm::format("%1$s%1$s", "foo");
+
     ConstevalFormatString<2>::Detail_CheckNumFormatSpecifiers("%2$s");
+    BOOST_CHECK_THROW(tfm::format("%2$s"), tfm::format_error);
+    BOOST_CHECK_THROW(tfm::format("%2$s", "foo"), tfm::format_error);
+    tfm::format("%2$s", "foo", "bar");
+
     ConstevalFormatString<2>::Detail_CheckNumFormatSpecifiers("%2$s 4$s %2$s");
-        ConstevalFormatString<129>::Detail_CheckNumFormatSpecifiers("%129$s 999$s %2$s");
+    BOOST_CHECK_THROW(tfm::format("%2$s 4$s %2$s"), tfm::format_error);
+    BOOST_CHECK_THROW(tfm::format("%2$s 4$s %2$s", "foo"), tfm::format_error);
+    tfm::format("%2$s 4$s %2$s", "foo", "bar");
+
+    ConstevalFormatString<12>::Detail_CheckNumFormatSpecifiers("%12$s 99$s %2$s");
+    BOOST_CHECK_THROW(tfm::format("%12$s 99$s %2$s"), tfm::format_error);
+    BOOST_CHECK_THROW(tfm::format("%12$s 99$s %2$s", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11"), tfm::format_error);
+    tfm::format("%12$s 99$s %2$s", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12");
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%02d");
+    BOOST_CHECK_THROW(tfm::format("%02d"), tfm::format_error);
+    tfm::format("%02d", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%+2s");
-        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%+2s");
+    BOOST_CHECK_THROW(tfm::format("%+2s"), tfm::format_error);
+    tfm::format("%+2s", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%.6i");
+    BOOST_CHECK_THROW(tfm::format("%.6i"), tfm::format_error);
+    tfm::format("%.6i", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%5.2f");
-        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%#x");
+    BOOST_CHECK_THROW(tfm::format("%5.2f"), tfm::format_error);
+    tfm::format("%5.2f", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%1$5i");
+    BOOST_CHECK_THROW(tfm::format("%1$5i"), tfm::format_error);
+    tfm::format("%1$5i", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%1$-5i");
+    BOOST_CHECK_THROW(tfm::format("%1$-5i"), tfm::format_error);
+    tfm::format("%1$-5i", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%1$.5i");
-        // tinyformat accepts almost any "type" spec, even '%', or '_', or '\n'.
+    BOOST_CHECK_THROW(tfm::format("%1$.5i"), tfm::format_error);
+    tfm::format("%1$.5i", 1);
+
+    ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%#x");
+    BOOST_CHECK_THROW(tfm::format("%#x"), tfm::format_error);
+    tfm::format("%#x", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%123%");
+    BOOST_CHECK_THROW(tfm::format("%123%"), tfm::format_error);
+    tfm::format("%123%", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%123%s");
+    BOOST_CHECK_THROW(tfm::format("%123%s"), tfm::format_error);
+    tfm::format("%123%s", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_");
+    BOOST_CHECK_THROW(tfm::format("%_"), tfm::format_error);
+    tfm::format("%_", 1);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%\n");
-        return true; // All checks above compiled and passed
-    }());
-    static_assert([] {
+    BOOST_CHECK_THROW(tfm::format("%\n"), tfm::format_error);
+    tfm::format("%\n", 1);
+}
+
 // The `*` specifier behavior is unsupported and can lead to runtime
 // errors when used in a ConstevalFormatString. Please refer to the
 // note in the ConstevalFormatString docs.
+BOOST_AUTO_TEST_CASE(ConstevalFormatString_IncorrectCounts)
+{
+    int a{}, b{}, c{};
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%*c");
+    BOOST_CHECK_THROW(tfm::format("%*c"), tfm::format_error);
+    BOOST_CHECK_THROW(tfm::format("%*c", a), tfm::format_error);
+    tfm::format("%*c", a, b);
+
     ConstevalFormatString<2>::Detail_CheckNumFormatSpecifiers("%2$*3$d");
+    BOOST_CHECK_THROW(tfm::format("%2$*3$d"), tfm::format_error);
+    BOOST_CHECK_THROW(tfm::format("%2$*3$d", a), tfm::format_error);
+    BOOST_CHECK_THROW(tfm::format("%2$*3$d", a, b), tfm::format_error);
+    tfm::format("%2$*3$d", a, b, c);
+
     ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%.*f");
-        return true; // All checks above compiled and passed
-    }());
+    BOOST_CHECK_THROW(tfm::format("%.*f"), tfm::format_error);
+    BOOST_CHECK_THROW(tfm::format("%.*f", a), tfm::format_error);
+    tfm::format("%.*f", a, b);
+}
 
-    // Negative checks at runtime
+// Executed at runtime as exceptions are not allowed at compile time.
+BOOST_AUTO_TEST_CASE(ConstevalFormatString_NegativeChecks)
+{
     using ErrType = const char*;
 
     auto check_mix{[](const ErrType& str) { return str == std::string_view{"Format specifiers must be all positional or all non-positional!"}; }};
     BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%s%1$s"), ErrType, check_mix);
 
-    auto check_num_spec{[](const ErrType& str) { return str == std::string_view{"Format specifier count must match the argument count!"}; }};
-    BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers(""), ErrType, check_num_spec);
-    BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%s"), ErrType, check_num_spec);
-    BOOST_CHECK_EXCEPTION(ConstevalFormatString<2>::Detail_CheckNumFormatSpecifiers("%s"), ErrType, check_num_spec);
-    BOOST_CHECK_EXCEPTION(ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%1$s"), ErrType, check_num_spec);
-    BOOST_CHECK_EXCEPTION(ConstevalFormatString<2>::Detail_CheckNumFormatSpecifiers("%1$s"), ErrType, check_num_spec);
-
     auto check_0_pos{[](const ErrType& str) { return str == std::string_view{"Positional format specifier must have position of at least 1"}; }};
     BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%$s"), ErrType, check_0_pos);
     BOOST_CHECK_EXCEPTION(ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%$"), ErrType, check_0_pos);
```

</details>

#### Changes compared to fa26462e95291652b4021d91b014655f678149e8

- Interlaced tinyformat calls to prove and document parity between compile time and runtime checks. (The purpose isn't to to test tinyformat). Now using the less verbose `BOOST_CHECK_THROW` for tinyformat instead of `BOOST_CHECK_EXCEPTION`.
- Breaking apart test function into multiple functions.
- Still dropping a few negative tests, since they are proven by positive tests, even if less straightforward without the free function.

<details>
<summary>

#### Runtime performance comparison (+0.3ms/+1.8% longer)

</summary>

fa26462e95291652b4021d91b014655f678149e8 + diff:
```
hyperfine --warmup 100 "ctest -j 20 --test-dir build -R util_string_tests"
Benchmark 1: ctest -j 20 --test-dir build -R util_string_tests
  Time (mean ± σ):      16.8 ms ±   0.2 ms    [User: 11.4 ms, System: 6.1 ms]
  Range (min … max):    16.2 ms …  17.8 ms    156 runs
```

fa26462e95291652b4021d91b014655f678149e8:
```
hyperfine --warmup 100 "ctest -j 20 --test-dir build -R util_string_tests"
Benchmark 1: ctest -j 20 --test-dir build -R util_string_tests
  Time (mean ± σ):      16.5 ms ±   0.2 ms    [User: 10.6 ms, System: 6.5 ms]
  Range (min … max):    16.0 ms …  17.4 ms    154 runs
```
</details>

hodlinator replied:
<details>
<summary>

#### Diff on top of fae8c25d07cf003df1470699df4ef475055bb885

</summary>

```diff
diff --git a/src/test/util_string_tests.cpp b/src/test/util_string_tests.cpp
index 4c61632fce..ed383e9dad 100644
--- a/src/test/util_string_tests.cpp
+++ b/src/test/util_string_tests.cpp
@@ -17,7 +17,53 @@ template <unsigned NumArgs>
 inline void PassFmt(util::ConstevalFormatString<NumArgs> fmt)
 {
     decltype(fmt)::Detail_CheckNumFormatSpecifiers(fmt.fmt); // This was run at compile-time, but run it again at run-time to avoid -Wunused.
+
+    // Prove parity with tinyformat
+    switch (NumArgs) {
+    case 0:
+        tfm::format(fmt.fmt);
+        break;
+    case 1:
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt), tfm::format_error);
+        tfm::format(fmt.fmt, "foo");
+        break;
+    case 2:
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt), tfm::format_error);
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt, "foo"), tfm::format_error);
+        tfm::format(fmt.fmt, "foo", "bar");
+        break;
+    case 12:
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt), tfm::format_error);
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt, "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11"), tfm::format_error);
+        tfm::format(fmt.fmt, "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12");
+        break;
+    }
 }
+
+template <unsigned WrongNumArgs, unsigned CorrectArgs>
+inline void PassFmtIncorrect(util::ConstevalFormatString<WrongNumArgs> fmt)
+{
+    decltype(fmt)::Detail_CheckNumFormatSpecifiers(fmt.fmt); // This was run at compile-time, but run it again at run-time to avoid -Wunused.
+
+    // Disprove parity with tinyformat
+    static_assert(WrongNumArgs != CorrectArgs);
+    int a{}, b{}, c{};
+    switch (CorrectArgs) {
+    case 2:
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt), tfm::format_error);
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt, a), tfm::format_error);
+        tfm::format(fmt.fmt, a, b);
+        break;
+    case 3:
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt), tfm::format_error);
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt, a), tfm::format_error);
+        BOOST_CHECK_THROW(tfm::format(fmt.fmt, a, b), tfm::format_error);
+        tfm::format(fmt.fmt, a, b, c);
+        break;
+    }
+}
+
+// Executed at runtime as exceptions are not allowed at compile time.
 template <unsigned WrongNumArgs>
 inline void FailFmtWithError(std::string_view wrong_fmt, std::string_view error)
 {
@@ -40,7 +86,7 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
     PassFmt<1>("%1$s%1$s");
     PassFmt<2>("%2$s");
     PassFmt<2>("%2$s 4$s %2$s");
-    PassFmt<129>("%129$s 999$s %2$s");
+    PassFmt<12>("%12$s 999$s %2$s");
     PassFmt<1>("%02d");
     PassFmt<1>("%+2s");
     PassFmt<1>("%.6i");
@@ -58,9 +104,9 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
     // The `*` specifier behavior is unsupported and can lead to runtime
     // errors when used in a ConstevalFormatString. Please refer to the
     // note in the ConstevalFormatString docs.
-    PassFmt<1>("%*c");
-    PassFmt<2>("%2$*3$d");
-    PassFmt<1>("%.*f");
+    PassFmtIncorrect<1, 2>("%*c");
+    PassFmtIncorrect<2, 3>("%2$*3$d");
+    PassFmtIncorrect<1, 2>("%.*f");
 
     auto check_mix{"Format specifiers must be all positional or all non-positional!"};
     FailFmtWithError<1>("%s%1$s", check_mix);
```
</details>

#### Changes compared to fae8c25d07cf003df1470699df4ef475055bb885

tinyformat calls to prove and document parity/non-parity between compile time and runtime checks. (The purpose isn't to to test tinyformat). (Using the less verbose `BOOST_CHECK_THROW` for tinyformat instead of `BOOST_CHECK_EXCEPTION`).

<details>
<summary>

#### Runtime performance comparison (+0.1ms/+0.6% longer)

</summary

fae8c25d07cf003df1470699df4ef475055bb885 + diff:
```
hyperfine --warmup 100 "ctest -j 20 --test-dir build -R util_string_tests"
Benchmark 1: ctest -j 20 --test-dir build -R util_string_tests
  Time (mean ± σ):      16.8 ms ±   0.4 ms    [User: 10.9 ms, System: 6.6 ms]
  Range (min … max):    16.4 ms …  19.9 ms    154 runs
```

fae8c25d07cf003df1470699df4ef475055bb885:
```
hyperfine --warmup 100 "ctest -j 20 --test-dir build -R util_string_tests"
Benchmark 1: ctest -j 20 --test-dir build -R util_string_tests
  Time (mean ± σ):      16.7 ms ±   0.3 ms    [User: 10.8 ms, System: 6.6 ms]
  Range (min … max):    16.2 ms …  17.8 ms    157 runs
```
</details>

hodlinator replied:
> > Still dropping a few negative tests, since they are proven by positive tests, even if less straightforward without the free function.
>
> I don't think this is true. Without the free function

True, forgot that that last condition was back in the non-free function. Negative tests are not removed in [latest diff](https://github.com/bitcoin/bitcoin/pull/30546#discussion_r1756443830).

l0rinc replied:
> Otherwise, it will never be ready because there is always a way to come up with an idea to improve something

I think we're getting closer, your patience is appreciated :)

hodlinator replied:
Yeah, I can do it as a follow-up since by now it does not really modify your existing code that much.

If you need to re-touch, I would appreciate if you just took this diff:
```diff
-    PassFmt<129>("%129$s 999$s %2$s");
+    PassFmt<12>("%12$s 999$s %2$s");
```
to make the follow-up cleaner.

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -0,0 +1,86 @@
+    PassFmt<1>("%1$-5i");
+    PassFmt<1>("%1$.5i");
+    // tinyformat accepts almost any "type" spec, even '%', or '_', or '\n'.
+    PassFmt<1>("%123%");
+    PassFmt<1>("%123%s");
```
This seems like an error to me, shouldn't trailing `%` be invalid? It fails for `"%123 %"`, shouldn't it fail for these as well?
The reason for the misclassification is https://github.com/bitcoin/bitcoin/pull/30546#discussion_r1754908585, i.e. that when detecting a `%` we skip the `if (++it >= str.end())` check.

The example in https://github.com/bitcoin/bitcoin/pull/30546/files#r1754643534 detects this as an error

l0rinc replied:
I have of course seen the comment and believe it, but I don't think we should mandate `"%123%"` being correct (i.e. not just allowing it, but adding test for it), while disallowing `"%123 %", "%" and "%1$".

l0rinc replied:
I understand that, but why are we making sure that the validation allows nonsensical values?

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -0,0 +1,86 @@
+template <unsigned WrongNumArgs>
+inline void FailFmtWithError(std::string_view wrong_fmt, std::string_view error)
+{
+    using ErrType = const char*;
+    auto check_throw{[error](const ErrType& str) { return str == error; }};
+    BOOST_CHECK_EXCEPTION(util::ConstevalFormatString<WrongNumArgs>::Detail_CheckNumFormatSpecifiers(wrong_fmt), ErrType, check_throw);
```
We could standardize this by extending `HasReason` in `setup_common.h`:
```C++
class HasReason
{
public:
    explicit HasReason(const std::string_view& reason) : m_reason(reason) {}
    bool operator()(std::string_view s) const { return s.find(m_reason) != std::string_view::npos; }
    bool operator()(const std::exception& e) const { return (*this)(e.what()); }

private:
    const std::string_view m_reason;
};
```
resulting in
```C++
BOOST_CHECK_EXCEPTION(util::ConstevalFormatString<WrongNumArgs>::Detail_CheckNumFormatSpecifiers(wrong_fmt), const char*, HasReason(error));
```

l0rinc replied:
Ok, though I doubt it will get merged if it's in a separate PR...

l0rinc replied:
Ok, looking forward to the follow-ups

l0rinc replied:
Pushed this change in https://github.com/bitcoin/bitcoin/pull/30921

l0rinc reviewed (changes requested):
It seems to me we still have a remaining trailing formatter bug that we should fix before merging

l0rinc reviewed (changes requested):
It seems to me we still have a remaining trailing formatter bug that we should fix before merging

stickies-v reviewed (approved):
ACK fa37d9b0fd0b8f1b23163887618b2fa1a817d596

I like the new test structure and added documentation, thanks! Left some non-blocking nits, only if you need to force-push.

stickies-v commented:
re-ACK fa5bc450d5d4c1d2daf7b205f1678402c3c21678
No changes except variable naming and small doc update.

hodlinator reviewed (approved):
ACK fa5bc450d5d4c1d2daf7b205f1678402c3c21678

`git range-diff master fae8c25 fa5bc45`

Only minor comment adjustments and clearer variable names since last reviewed commits.

Since GitHub is [troublesome](https://github.com/bitcoin/bitcoin/pull/30546#discussion_r1756561525), would prefer this change if you re-touch:
```diff
-    PassFmt<129>("%129$s 999$s %2$s");
+    PassFmt<12>("%12$s 999$s %2$s");
```
Makes it more reasonable to add comparison to tinyformat in follow-up with 12 args rather than 129, but still more than 1 digit.

l0rinc commented:
ACK fa5bc450d5d4c1d2daf7b205f1678402c3c21678

The change progressed a lot from its first version and based on the enthusiasm among reviewers I'm looking forward to continuing these in follow-up PRs.

ryanofsky reviewed (approved):
Code review ACK fa5bc450d5d4c1d2daf7b205f1678402c3c21678

Just test code cleanup since last review
